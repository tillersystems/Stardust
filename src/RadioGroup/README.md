# RadioGroup

### Usage

```jsx
import { RadioGroup } from '@tillersystems/stardust';
```

<!-- STORY -->

<!-- PROPS -->

### Example

```jsx
import { RadioGroup } from '@tillersystems/stardust';

render() {
  return (
    <RadioGroup groupName="vegetable">
      <RadioButton value="artichoke" id="artichoke" >artichoke</RadioButton>
      <RadioButton value="beetroot" id="beetroot">beetroot</RadioButton>
      <RadioButton value="pumpkin" id="pumpkin">pumpkin</RadioButton>
    </RadioGroup>
  );
}
```
