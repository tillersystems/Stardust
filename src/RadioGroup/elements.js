import styled, { css } from 'styled-components';

/**
 * Wrapper for the actual check box and its label.
 *
 * @return {jsx}
 */
export const Wrapper = styled.div`
  display: flex;
  align-items: baseline;
  flex-direction: column;

  padding: 0.6rem;

  ${({ isRow }) =>
    isRow &&
    css`
      flex-direction: row;
    `};

  ${({ disabled }) =>
    disabled &&
    css`
      cursor: not-allowed;
    `};
`;
