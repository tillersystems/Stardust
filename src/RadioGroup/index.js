import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { Wrapper } from './elements';

/**
 * A RadioGroup component groups radio buttons.
 *
 */
class RadioGroup extends PureComponent {
  /** Internal state. */
  constructor() {
    super();
    this.state = {
      selectedValue: null,
    };
  }

  /**
   * Handles mounting in component's lifecycle.
   */
  componentDidMount() {
    const { selectedValue } = this.props;
    /* eslint-disable react/no-did-mount-set-state */
    if (selectedValue !== null) {
      this.setState({ selectedValue: selectedValue });
    }
  }

  /**
   * Handles update in component's lifecycle.
   *
   * @param {Object} prevProps - The component's previous props.
   */
  componentDidUpdate(prevProps) {
    const { selectedValue } = this.props;

    if (selectedValue !== prevProps.selectedValue) {
      /* eslint-disable react/no-did-update-set-state */
      this.setState({ selectedValue });
    }
  }

  handleChange = value => {
    const { onChange } = this.props;

    this.setState(
      {
        selectedValue: value,
      },
      () => {
        onChange && onChange(value);
      },
    );
  };

  /**
   * Renders the component.
   *
   * @return {jsx}
   */
  render() {
    const { selectedValue } = this.state;
    const { children, className, groupName, isRow } = this.props;

    const radios = React.Children.map(children, radio =>
      React.cloneElement(radio, {
        onChange: this.handleChange,
        selectedValue,
        name: groupName,
      }),
    );
    return (
      <Wrapper isRow={isRow} className={className}>
        {radios}
      </Wrapper>
    );
  }
}

const { bool, func, node, string } = PropTypes;

/** Prop types. */
RadioGroup.propTypes = {
  children: node,
  className: string,
  groupName: string,
  isRow: bool,
  onChange: func,
  selectedValue: string,
};

/** Default props. */
RadioGroup.defaultProps = {
  children: null,
  className: '',
  groupName: null,
  isRow: false,
  onChange: null,
  selectedValue: null,
};

export default RadioGroup;
