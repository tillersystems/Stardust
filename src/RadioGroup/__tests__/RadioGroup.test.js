import React from 'react';
import { fireEvent } from 'react-testing-library';

import RadioGroup from '..';
import { RadioButton } from '../..';

describe('<RadioGroup />', () => {
  test('should render without a problem', () => {
    const { container } = render(
      <RadioGroup groupName="vegetable">
        <RadioButton value="artichoke" id="artichoke">
          artichoke
        </RadioButton>
        <RadioButton value="beetroot" id="beetroot">
          beetroot
        </RadioButton>
        <RadioButton value="pumpkin" id="pumpkin">
          pumpkin
        </RadioButton>
      </RadioGroup>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should should have selected value by default', () => {
    const { getByLabelText } = render(
      <RadioGroup groupName="vegetable" selectedValue="artichoke">
        <RadioButton value="artichoke" id="artichoke">
          artichoke
        </RadioButton>
        <RadioButton value="beetroot" id="beetroot">
          beetroot
        </RadioButton>
        <RadioButton value="pumpkin" id="pumpkin">
          pumpkin
        </RadioButton>
      </RadioGroup>,
    );

    const artichokeNode = getByLabelText('artichoke');
    const beetrootNode = getByLabelText('beetroot');
    const pumpkinNode = getByLabelText('beetroot');

    expect(artichokeNode).toHaveAttribute('checked');
    expect(beetrootNode).not.toHaveAttribute('checked');
    expect(pumpkinNode).not.toHaveAttribute('checked');
  });

  test('should select another value', () => {
    const { getByLabelText } = render(
      <RadioGroup groupName="vegetable" selectedValue="artichoke">
        <RadioButton value="artichoke" id="artichoke">
          artichoke
        </RadioButton>
        <RadioButton value="beetroot" id="beetroot">
          beetroot
        </RadioButton>
        <RadioButton value="pumpkin" id="pumpkin">
          pumpkin
        </RadioButton>
      </RadioGroup>,
    );

    let artichokeNode = getByLabelText('artichoke');
    let beetrootNode = getByLabelText('beetroot');
    let pumpkinNode = getByLabelText('pumpkin');

    expect(artichokeNode).toHaveAttribute('checked');
    expect(beetrootNode).not.toHaveAttribute('checked');
    expect(pumpkinNode).not.toHaveAttribute('checked');

    fireEvent.click(beetrootNode);

    artichokeNode = getByLabelText('artichoke');
    beetrootNode = getByLabelText('beetroot');
    pumpkinNode = getByLabelText('pumpkin');

    expect(artichokeNode).not.toHaveAttribute('checked');
    expect(beetrootNode).toHaveAttribute('checked');
    expect(pumpkinNode).not.toHaveAttribute('checked');
  });

  test('should have radio button disabled', () => {
    const { getByLabelText } = render(
      <RadioGroup groupName="vegetable" selectedValue="artichoke">
        <RadioButton value="artichoke" id="artichoke" disabled>
          artichoke
        </RadioButton>
        <RadioButton value="beetroot" id="beetroot">
          beetroot
        </RadioButton>
        <RadioButton value="pumpkin" id="pumpkin">
          pumpkin
        </RadioButton>
      </RadioGroup>,
    );

    const artichokeNode = getByLabelText('artichoke');
    const beetrootNode = getByLabelText('beetroot');
    const pumpkinNode = getByLabelText('pumpkin');

    expect(artichokeNode).toHaveAttribute('disabled');
    expect(beetrootNode).not.toHaveAttribute('disabled');
    expect(pumpkinNode).not.toHaveAttribute('disabled');
  });

  test('should render as row', () => {
    const render = shallowWithTheme(
      <RadioGroup groupName="vegetable" isRow>
        <RadioButton value="artichoke" id="artichoke">
          artichoke
        </RadioButton>
        <RadioButton value="beetroot" id="beetroot">
          beetroot
        </RadioButton>
        <RadioButton value="pumpkin" id="pumpkin">
          pumpkin
        </RadioButton>
      </RadioGroup>,
    );
    expect(render.dive()).toMatchSnapshot();
  });

  test('should call componentDidUpdate when props are updated', () => {
    let selectedValue = 'artichoke';
    const { getByLabelText, rerender } = render(
      <RadioGroup groupName="vegetable" selectedValue={selectedValue}>
        <RadioButton value="artichoke" id="artichoke">
          artichoke
        </RadioButton>
        <RadioButton value="beetroot" id="beetroot">
          beetroot
        </RadioButton>
        <RadioButton value="pumpkin" id="pumpkin">
          pumpkin
        </RadioButton>
      </RadioGroup>,
    );
    let artichokeNode = getByLabelText('artichoke');
    let beetrootNode = getByLabelText('beetroot');
    let pumpkinNode = getByLabelText('pumpkin');

    expect(artichokeNode).toHaveAttribute('checked');
    expect(beetrootNode).not.toHaveAttribute('checked');
    expect(pumpkinNode).not.toHaveAttribute('checked');

    selectedValue = 'beetroot';
    rerender(
      <RadioGroup groupName="vegetable" selectedValue={selectedValue}>
        <RadioButton value="artichoke" id="artichoke">
          artichoke
        </RadioButton>
        <RadioButton value="beetroot" id="beetroot">
          beetroot
        </RadioButton>
        <RadioButton value="pumpkin" id="pumpkin">
          pumpkin
        </RadioButton>
      </RadioGroup>,
    );

    artichokeNode = getByLabelText('artichoke');
    beetrootNode = getByLabelText('beetroot');
    pumpkinNode = getByLabelText('pumpkin');

    expect(artichokeNode).not.toHaveAttribute('checked');
    expect(beetrootNode).toHaveAttribute('checked');
    expect(pumpkinNode).not.toHaveAttribute('checked');
  });
});
