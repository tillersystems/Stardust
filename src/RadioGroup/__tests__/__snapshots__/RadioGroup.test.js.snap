// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> should render as row 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.6rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  padding: 0.6rem;
}

.c1 > input {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1.6rem;
  height: 1.6rem;
  margin-right: 1.2rem;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 50%;
  background: linear-gradient( 0deg,hsl(240,33%,99%) 0%,hsl(0,100%,100%) 100% );
  box-shadow: inset 0 2px 0 0 hsla(0,100%,100%,0.05),0 1px 1px 0 hsl(240,33%,99%);
  -webkit-transition: background-color ease 0.25s;
  transition: background-color ease 0.25s;
}

.c3.checked {
  border: 1px solid #1873a0;
  background: hsl(200,74%,46%) linear-gradient( 0deg,hsla(0,100%,100%,0) 0%,hsla(0,100%,100%,0.1) 100% );
  box-shadow: inset 0 0.2rem 0 0 hsla(0,100%,100%,0.1);
  -webkit-transition: border ease 0.25s;
  transition: border ease 0.25s;
  position: relative;
}

.c3.checked:after {
  content: '';
  position: absolute;
  left: 50%;
  top: 50%;
  width: 0.6rem;
  height: 0.6rem;
  -webkit-transform: translate3d(-50%,-50%,0);
  -ms-transform: translate3d(-50%,-50%,0);
  transform: translate3d(-50%,-50%,0);
  background-color: hsl(0,100%,100%);
  border-radius: 50%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  color: hsl(207,13%,45%);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2"
    >
      <input
        id="artichoke"
        name="vegetable"
        type="radio"
        value="artichoke"
      />
      <div
        class="c3"
        data-testid="styled-radio"
      />
      artichoke
    </label>
  </div>
  <div
    class="c1"
  >
    <label
      class="c2"
    >
      <input
        id="beetroot"
        name="vegetable"
        type="radio"
        value="beetroot"
      />
      <div
        class="c3"
        data-testid="styled-radio"
      />
      beetroot
    </label>
  </div>
  <div
    class="c1"
  >
    <label
      class="c2"
    >
      <input
        id="pumpkin"
        name="vegetable"
        type="radio"
        value="pumpkin"
      />
      <div
        class="c3"
        data-testid="styled-radio"
      />
      pumpkin
    </label>
  </div>
</div>
`;

exports[`<RadioGroup /> should render without a problem 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  padding: 0.6rem;
}

.c1 > input {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1.6rem;
  height: 1.6rem;
  margin-right: 1.2rem;
  border: 1px solid hsl(207,22%,90%);
  border-radius: 50%;
  background: linear-gradient( 0deg,hsl(240,33%,99%) 0%,hsl(0,100%,100%) 100% );
  box-shadow: inset 0 2px 0 0 hsla(0,100%,100%,0.05),0 1px 1px 0 hsl(240,33%,99%);
  -webkit-transition: background-color ease 0.25s;
  transition: background-color ease 0.25s;
}

.c3.checked {
  border: 1px solid #1873a0;
  background: hsl(200,74%,46%) linear-gradient( 0deg,hsla(0,100%,100%,0) 0%,hsla(0,100%,100%,0.1) 100% );
  box-shadow: inset 0 0.2rem 0 0 hsla(0,100%,100%,0.1);
  -webkit-transition: border ease 0.25s;
  transition: border ease 0.25s;
  position: relative;
}

.c3.checked:after {
  content: '';
  position: absolute;
  left: 50%;
  top: 50%;
  width: 0.6rem;
  height: 0.6rem;
  -webkit-transform: translate3d(-50%,-50%,0);
  -ms-transform: translate3d(-50%,-50%,0);
  transform: translate3d(-50%,-50%,0);
  background-color: hsl(0,100%,100%);
  border-radius: 50%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  color: hsl(207,13%,45%);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2"
    >
      <input
        id="artichoke"
        name="vegetable"
        type="radio"
        value="artichoke"
      />
      <div
        class="c3"
        data-testid="styled-radio"
      />
      artichoke
    </label>
  </div>
  <div
    class="c1"
  >
    <label
      class="c2"
    >
      <input
        id="beetroot"
        name="vegetable"
        type="radio"
        value="beetroot"
      />
      <div
        class="c3"
        data-testid="styled-radio"
      />
      beetroot
    </label>
  </div>
  <div
    class="c1"
  >
    <label
      class="c2"
    >
      <input
        id="pumpkin"
        name="vegetable"
        type="radio"
        value="pumpkin"
      />
      <div
        class="c3"
        data-testid="styled-radio"
      />
      pumpkin
    </label>
  </div>
</div>
`;
